racingcar/game.go

```diff
@@ -0,0 +1,98 @@
+package racingcar
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	minParticipants     = 1
+	maxParticipants     = 20
+	allowedMinTurnCount = 1
+	allowedMaxTurnCount = 100
+)
+
+type RaceCars []Car
+
+type Game struct {
+	Cars             RaceCars
+	maxTurnCount     int
+	currentTurnCount int
+}
+
+func NewGame(carNames []string, maxTurnCount int) (*Game, error) {
+	cars, err := initCars(carNames)
+	if err != nil {
+		return nil, err
+	}
+
+	if !validateGameCount(maxTurnCount) {
+		return nil, fmt.Errorf("시도 횟수는 %d이상 %d이하의 숫자만 가능합니다.", allowedMinTurnCount, allowedMaxTurnCount)
+	}
+
+	return &Game{
+		Cars:             cars,
+		maxTurnCount:     maxTurnCount,
+		currentTurnCount: 0,
+	}, nil
+}
+
+func initCars(carNames []string) ([]Car, error) {
+	if !validateNumberOfCar(len(carNames)) {
+		return nil, fmt.Errorf("자동차의 수는 %d대 이상 %d대 이하만 가능합니다.", minParticipants, maxParticipants)
+	}
+
+	cars := make([]Car, len(carNames))
+	for i, name := range carNames {
+		car, err := NewCar(strings.TrimSpace(name))
+		if err != nil {
+			return nil, err
+		}
+		cars[i] = *car
+	}
+
+	return cars, nil
+}
+
+func validateNumberOfCar(numberOfCar int) bool {
+	return numberOfCar >= minParticipants && numberOfCar <= maxParticipants
+}
+
+func validateGameCount(count int) bool {
+	return count >= allowedMinTurnCount && count <= allowedMaxTurnCount
+}
+
+func (game *Game) IsEnd() bool {
+	return game.currentTurnCount == game.maxTurnCount
+}
+
+func (game *Game) PlayTurn(moveStrategy MoveStrategy) error {
+	if game.IsEnd() {
+		return fmt.Errorf("게임이 종료되었습니다.")
+	}
+	for i := range game.Cars {
+		PlayTurn(&game.Cars[i], moveStrategy)
+	}
+	game.currentTurnCount++
+	return nil
+}
+
+func PlayTurn(car *Car, isMovable MoveStrategy) {
+	if isMovable() {
+		car.MoveForward()
+	}
+}
```

Comment:
P2. 저의 경우 Cars를 `[]*Car` 타입으로 관리하고 아래와 같이 구현할 것 같아요.

P4. 아래와 같이 함수 타입을 그대로 매개변수 타입으로 쓰는 경우도 흔합니다.

```suggestion
func (game *Game) PlayTurn(moveStrategy func() bool) error {
	if game.IsEnd() {
		return fmt.Errorf("게임이 종료되었습니다.")
	}
	for _, car := range game.Cars {
		if moveStrategy() {
			car.MoveForward()
		}
	}
	game.currentTurnCount++
	return nil
}
```

Reply:
팁 주신 부분들 감사합니다!!